### ISSUE TITLE
ScreenRanking doesn't get updated after ITG or FA+ mode gameplay

### ISSUE THREAD

**Issue by NathanRichards007** [dguzek/Simply-Love-SM5#256]
For some reason, after I used NameEntry screen on ITG or FA+ mode, submit the 4-letter/number name and then when I go to the Rankings screen, it doesn't get updated, but it shows the songs that I've more recently played.
![Screenshot 2021-02-10 193837](https://user-images.githubusercontent.com/78886988/107591014-b0021800-6bd7-11eb-8836-3174c5adb533.png)

**Comment by quietly-turning** [comment]
This can occur if data from Casual-Stats.xml is currently loaded in memory, which happens when Casual Mode was most recently played.

Casual Mode doesn't save HighScore data (score, date, name, etc.) to the Machine Profile.  This was by design.

* I've never seen a player that Casual Mode was designed for *want* to enter a high score name; they frequently did not even finished their allotted set (e.g. 2 songs on the public machine I previously managed)
* At the time, I saw it as an opportunity to have fewer disk write operations, and a good thing.

SL's modes were introduced [quite a long time ago](https://github.com/Simply-Love/Simply-Love-SM5/releases/tag/3.0).  I'm still strongly in favor of the existence of Casual Mode as a UI path, but having multiple Stats.xml files (for Casual, for ITG, for FA+, previously for StomperZ) seems much less good now.

I can provide more historical context and discuss issues with the multiple Stats files approach, but it gets outside the scope of this issue pretty quickly.

**Comment by NathanRichards007** [comment]
Is there a solution though to fix it?

**Comment by quietly-turning** [comment]
For now, a player-facing solution is "don't play Casual Mode," though that's more workaround than solution.

The fix, long-term, is for Simply Love to move away from relying on multiple Stats.xml files, but

* I haven't fully thought through implementation details.
* A backwards-compatible migration path for the community would be complicated to implement and a support nightmare.
* I'm probably the only person that presently has a deep enough working knowledge of Simply Love's nooks and crannies to immediately undertake this, and I cannot take on the responsibility right now or in the foreseeable future.

**Comment by NathanRichards007** [comment]
At that point, I was actually playing on itg mode, and although I haven't had the casual mode stats file/haven't actually played casual mode; as a precaution, I've deleted the machineprofiles folder off of the installation folder.

**Comment by quietly-turning** [comment]
After a night's sleep, I better appreciate that my responses above are confusing.  I was trying to convey the complexity of the what I thought the problem was to the current maintainers of Simply-Love-SM5, and wasn't very helpful to you, @NathanRichards007.  Sorry about this.

**Comment by NathanRichards007** [comment]
I meant although I haven't played casual mode; I still don't have the high score screen updated after I put a 4 letter name on the name entry screen on ITG mode or FA+ mode.

**Comment by quietly-turning** [comment]
Are you seeing those HighScores appear in ScreenSelectMusic, like this?

![Screen Shot 2021-02-12 at 1 39 15 PM](https://user-images.githubusercontent.com/1253483/107808802-2453cd00-6d38-11eb-9a41-9144a04e7d70.png)

**Comment by NathanRichards007** [comment]
> Are you seeing those HighScores appear in ScreenSelectMusic, like this?
> 
> ![Screen Shot 2021-02-12 at 1 39 15 PM](https://user-images.githubusercontent.com/1253483/107808802-2453cd00-6d38-11eb-9a41-9144a04e7d70.png)

Yes; but on the ranking screen, I don't see the high scores; its blank like I've played casual mode, although I didn't play that mode.

**Comment by quietly-turning** [comment]
Thanks for responding and providing more information. Hopefully this helps someone debug this.

**Comment by quietly-turning** [comment]
Other information you can provide includes:

1. What version of SM5 are you using?
2. What OS are you using?
3. Where do you have SM5 installed?
4. What version of Simply-Love-SM5 are you using?

Finally, if you temporarily enable `ShowThemeErrors` and play through a game cycle, and then visit ScreenRankingSingle and ScreenRankingDouble, are any Lua errors logged at any point?  They would appear at the top of the screen in a green box.  If yes, please attach a screenshot.

You can toggle `ShowThemeErrors` on/off via
1. hold <kbd>F3</kbd> and tap <kbd>F6</kbd> , then
2. hold <kbd>F3</kbd> and tap <kbd>8</kbd>

**Comment by NathanRichards007** [comment]
> Other information you can provide includes:
> 
> 1. What version of SM5 are you using?
> 2. What OS are you using?
> 3. Where do you have SM5 installed?
> 4. What version of Simply-Love-SM5 are you using?
> 
> Finally, if you temporarily enable `ShowThemeErrors` and play through a game cycle, and then visit ScreenRankingSingle and ScreenRankingDouble, are any Lua errors logged at any point? They would appear at the top of the screen in a green box. If yes, please attach a screenshot.
> 
> You can toggle `ShowThemeErrors` on/off via
> 
> 1. hold F3 and tap F6 , then
> 2. hold F3 and tap 8

1. Stepmania 5.1 beta 2
2. Windows 10 1909
3. In my external hard drive (using the portable.ini for the sm5.1 installation folder)
4. Simply-Love-SM5 4.9.1 (the latest version)

And, for some reason; it didn't give me an error on the high score ranking screen. So I didn't give a screenshot.

**Comment by quietly-turning** [comment]
If you're able to, create a separate SM5 install that is not on an external drive and doesn't use portable.ini (installing to `C:\Games\` is [recommended](https://github.com/Simply-Love/Simply-Love-SM5/blob/7d876a41c5dc10fde40b9adefc1f860576aaa470/Other/Documentation/TroubleshootingStepMania-README.md#windows)), play a game session with that, and report back.

**Comment by NathanRichards007** [comment]
No errors, but surprisingly after playing the same song (in this case, Goin' Under) for a few times; the profile record appears on the high score screen. Maybe it might be the maximum machine scores & player scores default/recommended settings on the operator options screen that might be the issue.
![Screenshot 2021-02-12 155506](https://user-images.githubusercontent.com/78886988/107821688-e613d900-6d4a-11eb-8a22-dd59ac921aa0.png)

**Comment by quietly-turning** [comment]
I probably should have mentioned earlier that ScreenRankingSingle and ScreenRankingDouble only show HighScores for [Hard and Challenge stepcharts](https://quietly-turning.github.io/Lua-For-SM5/LuaAPI#Enums-Difficulty).  HighScores on Beginner, Easy, Medium, and Edit stepcharts will not appear.  Is that possibly the issue here?

**Comment by NathanRichards007** [comment]
No, since I was playing on the hard difficulty at that time; maybe for some reason, it won't show the first place original hard or challenge scoring until playing the same song on either of the two last difficulty for the second time and getting a better score.

**Comment by quietly-turning** [comment]
Thanks for providing all the information.   I'm confident it will help someone get up and running with this faster than without.

**Comment by NathanRichards007** [comment]
no problem; on the default sm5 theme, the scoring on first place appears on the ranking screen; I hope you sort of investigate this issue and possibly patch it if it is fixable.
